lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+  // associated domain, it must be identical. Otherwise we'll have to report the
+  // conflicting domains to the user.
+  auto &entries = domains[module];
+  if (entries.empty() || llvm::all_of(entries, [&](const auto &entry) {
+        return entry.first != domain;
+      }))
+    entries.push_back({domain, instPath});
+
+  // Traverse the child instances.
+  InstancePath childPath = instPath;
+  for (auto record : instGraph[module]->instances()) {
+    auto submodule = dyn_cast<FModuleOp>(record->getTarget()->getModule());
+    if (!submodule)
+      continue;
+    childPath.push_back(record->getInstance());
+    buildDomains(submodule, childPath, domain.reset, instGraph, indent + 1);
+    childPath.pop_back();
+  }
+}
+
+/// Determine how the reset for each module shall be implemented.
+void InferResetsPass::determineImpl() {
+  for (auto &it : domains) {
+    auto module = cast<FModuleOp>(it.first);
+    auto &domain = it.second.back().first;
+    determineImpl(module, domain);
+  }
+}
+
+/// Determine how the reset for a module shall be implemented. This function
+/// fills in the `existingValue`, `existingPort`, and `newPortName` fields of
+/// the given reset domain.
+///
+/// Generally it does the following:
+/// - If the domain has explicitly no reset ("ignore"), leaves everything empty.
+/// - If the domain is the place where the reset is defined ("top"), fills in
+///   the existing port/wire/node as reset.
+/// - If the module already has a port with the reset's name:
+///   - If the type is `asyncreset`, reuses that port.
+///   - Otherwise appends a `_N` suffix with increasing N to create a yet-unused
+///     port name, and marks that as to be created.
+/// - Otherwise indicates that a port with the reset's name should be created.
+///
+void InferResetsPass::determineImpl(FModuleOp module, ResetDomain &domain) {
+  if (!domain.reset)
+    return; // nothing to do if the module needs no reset
+  LLVM_DEBUG(llvm::dbgs() << "Planning reset for " << module.getName() << "\n");
+
+  // If this is the root of a reset domain, we don't need to add any ports
+  // and can just simply reuse the existing values.
+  if (domain.isTop) {
+    LLVM_DEBUG(llvm::dbgs() << "- Rooting at local value "
+                            << getResetName(domain.reset) << "\n");
+    domain.existingValue = domain.reset;
+    if (auto blockArg = domain.reset.dyn_cast<BlockArgument>())
+      domain.existingPort = blockArg.getArgNumber();
+    return;
+  }
+
+  // Otherwise, check if a port with this name and type already exists and
+  // reuse that where possible.
+  auto neededName = getResetName(domain.reset);
+  auto neededType = domain.reset.getType();
+  LLVM_DEBUG(llvm::dbgs() << "- Looking for existing port " << neededName
+                          << "\n");
+  auto portNames = getModulePortNames(module);
+  auto ports = llvm::enumerate(llvm::zip(portNames, module.getArguments()));
```

Comment:
I'd personally drop the enumerate here.  I think you can use `domain.existingPort = portIt-ports.begin()` to get the index since these are random access iterators.

Reply:
That's a nice idea. Changed this around.